src/ott/neural/models/base_solver.py

```diff
@@ -0,0 +1,485 @@
+# Copyright OTT-JAX
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import abc
+import pathlib
+import types
+from typing import Any, Callable, Dict, Literal, Mapping, Optional, Tuple, Union
+
+import jax
+import jax.numpy as jnp
+
+import optax
+from flax.training import train_state
+
+from ott.geometry import costs, pointcloud
+from ott.problems.linear import linear_problem
+from ott.problems.quadratic import quadratic_problem
+from ott.solvers.linear import sinkhorn
+
+__all__ = ["BaseNeuralSolver", "ResampleMixin", "UnbalancednessMixin"]
+
+
+class BaseNeuralSolver(abc.ABC):
+  """Base class for neural solvers.
+
+  Args:
+    iterations: Number of iterations to train for.
+    valid_freq: Frequency at which to run validation.
+  """
+
+  def __init__(self, iterations: int, valid_freq: int, **_: Any):
+    self.iterations = iterations
+    self.valid_freq = valid_freq
+
+  @abc.abstractmethod
+  def setup(self, *args: Any, **kwargs: Any):
+    """Setup the model."""
+
+  @abc.abstractmethod
+  def __call__(self, *args: Any, **kwargs: Any):
+    """Train the model."""
+
+  @abc.abstractmethod
+  def transport(self, *args: Any, forward: bool, **kwargs: Any) -> Any:
+    """Transport."""
+
+  @abc.abstractmethod
+  def save(self, path: pathlib.Path):
+    """Save the model."""
+
+  @abc.abstractmethod
+  def load(self, path: pathlib.Path):
+    """Load the model."""
+
+  @property
+  @abc.abstractmethod
+  def training_logs(self) -> Dict[str, Any]:
+    """Return the training logs."""
+
+
+class ResampleMixin:
+  """Mixin class for mini-batch OT in neural optimal transport solvers."""
+
+  def _resample_data(
+      self,
+      key: jax.random.KeyArray,
+      tmat: jnp.ndarray,
+      source_arrays: Tuple[jnp.ndarray, ...],
+      target_arrays: Tuple[jnp.ndarray, ...],
+  ) -> Tuple[jnp.ndarray, ...]:
+    """Resample a batch according to coupling `tmat`."""
+    tmat_flattened = tmat.flatten()
+    indices = jax.random.choice(key, len(tmat_flattened), shape=[tmat.shape[0]])
+    indices_source = indices // tmat.shape[1]
+    indices_target = indices % tmat.shape[1]
+    return tuple(
+        b[indices_source, :] if b is not None else None for b in source_arrays
```

Comment:
I think this can be done in a more general fashion using `jax.tree_util.tree_map`.